### PR TITLE
CTCCriterion variable length and batchFirst support

### PR DESCRIPTION
**Pull request by SeanNaren**
I've modified the criterion to support variable length sequences in a batch and has a batchFirst param for inputs of ```batch x seqLength x inputDim```. Let me know of any feedback etc and thanks @willfrey, @seed93 for your help!

Side note, if anyone is using the old criterion like before you will need to pass the sizes in. The previous criterion assumed all sizes were the same, so to make it work just calculate your sizes via your inputs like this:

```
local tensorSizes = outputs:size()
local sizes = torch.Tensor(tensorSizes[1]):fill(tensorSizes[2])
local loss = criterion:forward(outputs, targets, sizes)
```

### CODE REVIEW

**Comment by seed93** [comment]
Great!

**Comment by soumith** [comment]
Thanks! :)

**Comment by seed93** [comment]
It seems that loss and grad is sum of all samples in a minibatch, should we divide them by batchsize?

**Comment by SeanNaren** [comment]
@seed93 I think you're right, just had a look at the warp-ctc paper and their costs were definitely divided by batch size. I'll open a PR as soon as I can :)

**Comment by seed93** [comment]
acts is not initialized, assertion failed when given 2 dimensions.
